pkg/controller/upgradeconfig/cluster_upgrader.go

```diff
@@ -0,0 +1,768 @@
+	}
+
+	return true, nil
+}
+
+// TODO
+func RemoveMaintWindow(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	return true, nil
+}
+
+// This perform cluster health check after upgrade
+func PostClusterHealthCheck(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	return performClusterHealthCheck(c)
+}
+
+// Check whether nodes are upgraded or not
+func NodesUpgraded(c client.Client, nodeType string, reqLogger logr.Logger) (bool, error) {
+	configPool := &machineconfigapi.MachineConfigPool{}
+	err := c.Get(context.TODO(), types.NamespacedName{Name: nodeType}, configPool)
+	if err != nil {
+		return false, nil
+	}
+	//TODO send timeout alert if wait timeout
+	if configPool.Status.ReadyMachineCount != configPool.Status.UpdatedMachineCount {
+		errMsg := fmt.Sprintf("not all %s are upgraded, upgraded: %v, ready: %v", nodeType, configPool.Status.UpdatedMachineCount, configPool.Status.ReadyMachineCount)
+		reqLogger.Info(errMsg)
+		return false, fmt.Errorf(errMsg)
+	}
+
+	return true, nil
+}
+
+// This check whether control plane is upgraded or not
+func ControlPlaneUpgraded(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	clusterVersion := &configv1.ClusterVersion{}
+
+	err := c.Get(context.TODO(), types.NamespacedName{Name: "version"}, clusterVersion)
+	if err != nil {
+		return false, err
+	}
+	for _, c := range clusterVersion.Status.History {
+		if c.State == configv1.CompletedUpdate && c.Version == upgradeConfig.Spec.Desired.Version {
+			return true, nil
+		}
+	}
+	return false, nil
+
+}
+
+// This trigger the upgrade process
+func (c ClusterUpgrader) UpgradeCluster(cl client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) error {
+
+	log.Info("upgrading cluster")
+	history := upgradeConfig.Status.History.GetHistory(upgradeConfig.Spec.Desired.Version)
+	conditions := history.Conditions
+
+	history.Phase = upgradev1alpha1.UpgradePhaseUpgrading
+	history.StartTime = &metav1.Time{Time: time.Now()}
+	upgradeConfig.Status.History.SetHistory(*history)
+	err := cl.Status().Update(context.TODO(), upgradeConfig)
+	if err != nil {
+		log.Error(err, "failed to update upgradeconfig")
+	}
+	for _, key := range Ordering() {
+
+		log.Info(fmt.Sprintf("Perform %s", key))
+
+		condition := conditions.GetCondition(key)
+		if condition == nil {
+			reqLogger.Info(fmt.Sprintf("Adding %s condition", key))
+			condition = newUpgradeCondition(fmt.Sprintf("start %s", key), fmt.Sprintf("start %s", key), key, corev1.ConditionFalse)
+			condition.StartTime = &metav1.Time{time.Now()}
+			conditions.SetCondition(*condition)
+			history.Conditions = conditions
+			upgradeConfig.Status.History.SetHistory(*history)
+			err := cl.Status().Update(context.TODO(), upgradeConfig)
+			if err != nil {
+				return err
+			}
+		}
+		if condition.Status == corev1.ConditionTrue {
```

Comment:
We should reconcile on each stage every (state should be based on the cluster state not our status) so I would remove this

Reply:
We are doing step by step upgrade, when we finished step 1, and running step2, and when the 2nd time reconcile starts, we cannot go back check whether step1 finished or not based on the current cluster status, that's why we so much rely on the `conditions` here.

Reply:
In my experience this always ends up being a problem eventually. If an upgrade goes wrong and a sre needs to restart the upgrade from a certain point this can stop the operator from recovering correctly (bar manually updating the status)

Reply:
It's a lot more work than expected without using this, let's address this problem in a follow up jira, I will create one and linked here

Reply:
https://issues.redhat.com/browse/OSD-3916